### PR TITLE
OSDOCS-3040: Check for config drift regularly

### PR DESCRIPTION
**Pull request by mburke5678**
https://issues.redhat.com/browse/OSDOCS-3040

[Understanding the Config Drift Monitor](https://deploy-preview-40133--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/machine-configuration-tasks.html#machine-config-drift-detection_post-install-machine-configuration-tasks)
New sesction 

[Machine config overview](https://deploy-preview-40133--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/machine-configuration-tasks.html#machine-config-overview-post-install-machine-configuration-tasks)
Last paragraph and Additional Resources

[Understanding the Machine Config Operator](https://deploy-preview-40133--osdocs.netlify.app/openshift-enterprise/latest/architecture/control-plane.html#understanding-machine-config-operator_control-plane)
Last paragraph and Additional Resources. Same paragraph and link.

[Node configuration management with machine config pools](https://deploy-preview-40133--osdocs.netlify.app/openshift-enterprise/latest/architecture/control-plane#architecture-machine-config-pools_control-plane)
Last paragraph and Additional Resources Same paragraph and link.

[Using MachineConfig objects to configure nodes](https://deploy-preview-40133--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/machine-configuration-tasks.html#using-machineconfigs-to-change-machines)
|ast paragraph with link Same paragraph.

### CODE REVIEW

**Comment by mburke5678** [comment]
@cheesesashimi Can you please take a look at my PR?

File: post_installation_configuration/machine-configuration-tasks.adoc [hunk 1]
```diff
@@ -26,6 +27,8 @@ link:https://access.redhat.com/solutions/4510281[updating] SSH authorized keys,
 
 {product-title} supports link:https://coreos.github.io/ignition/configuration-v3_2/[Ignition specification version 3.2]. All new machine configs you create going forward should be based on Ignition specification version 3.2. If you are upgrading your {product-title} cluster, any existing Ignition specification version 2.x machine configs will be translated automatically to specification version 3.2.
 
+There might be situations where the configuration on a node does not match fully match what the currently-applied machine config specifies. This state is called configuration drift. The Machine Config Daemon regularly checks the node configuration for configuration drift. If the MCD detects configuration drift, the MCO disables the node until an admin corrects the node configuration. For more information, see xref:../post_installation_configuration/machine-configuration-tasks.adoc#machine-config-drift-detection_post-install-machine-configuration-tasks[Understanding configuration drift detection].
```
Comment (cheesesashimi) [inline post_installation_configuration/machine-configuration-tasks.adoc:30]: I noticed this sentence used throughout the documentation updates, but I'll leave my comment in this occurrence (though it applies to all of them):

> If the MCD detects configuration drift, the MCO disables the node until an admin corrects the node configuration.

This implies (to me, anyway) that the node is no longer operational or available. When a node is degraded, it means that the machine configuration cannot be updated. In the case of config drift, it means the cluster admin must take action to restore the node to a known-good state. The node is still otherwise online and operational, with no workload disruption.

It would be more accurate to say, "the MCO marks the node degraded until..." (or similar).

File: modules/machine-config-drift-detection.adoc [hunk 2]
```diff
@@ -0,0 +1,106 @@
+// Module included in the following assemblies:
+//
+// * post_installation_configuration/machine-configuration-tasks.adoc
+
+[id="machine-config-drift-detection_{context}"]
+= Understanding configuration drift detection
+
+There might be situations when the on-disk state of a node differs from what is configured in the machine config. This is known as _configuration drift_. For example, a cluster admin might manually modify a file, a systemd unit file, file permissions or other operating system feature that was configured through a machine config. This causes configuration drift. Configuration drift can cause problems between nodes in a Machine Config Pool or when the machine configs are updated.
```
Comment (cheesesashimi) [inline modules/machine-config-drift-detection.adoc:8]: > For example, a cluster admin might manually modify a file, a systemd unit file, file permissions or other operating system feature that was configured through a machine config.

It would be more accurate to omit the, "other operating system feature" in this case. We only detect config drift based upon file contents, systemd unit file contents, or file permissions. Other operating system features such as kernel arguments, SSH keys, etc. are not checked.

**Review by cheesesashimi — COMMENTED** [review]
Thanks for the doc updates. They look great! I only have two specific comments.

**Comment by mburke5678** [comment]
@rioliu-rh PTAL. Thank you!

**Comment by rioliu-rh** [comment]
looks good to me, waiting for Sergio's review

**Review by sergiordlr — APPROVED** [review]
Looks good to me.

Thank you very much!

**Comment by mburke5678** [comment]
@cheesesashimi @rioliu-rh @sergiordlr Thank you for your help!

File: modules/architecture-machine-config-pools.adoc [hunk 3]
```diff
@@ -24,3 +24,6 @@ Any node labeled with the `infra` role that is only running infra workloads is n
 ====
 
 The MCO applies updates for pools independently; for example, if there is an update that affects all pools, nodes from each pool update in parallel with each other. If you add a custom pool, nodes from that pool also attempt to update concurrently with the master and worker nodes.
+
+There might be situations where the configuration on a node does not fully match what the currently-applied machine config specifies. This state is called _configuration drift_. The Machine Config Daemon (MCD) regularly checks the nodes for configuration drift. If the MCD detects configuration drift, the MCO marks the node `degraded` until an admin corrects the node configuration. A degraded node is online and operational. However, it cannot be updated.
```
Comment (bergerhoffer) [inline modules/architecture-machine-config-pools.adoc:28]: * admin -> administrator
* Not critical, but I'd consider combining the last two sentences, that it's online/operational, but can't be updated

Same for several other places that use this/similar wording

File: modules/machine-config-drift-detection.adoc [hunk 4]
```diff
@@ -0,0 +1,106 @@
+// Module included in the following assemblies:
+//
+// * post_installation_configuration/machine-configuration-tasks.adoc
+
+[id="machine-config-drift-detection_{context}"]
+= Understanding configuration drift detection
+
+There might be situations when the on-disk state of a node differs from what is configured in the machine config. This is known as _configuration drift_. For example, a cluster admin might manually modify a file, a systemd unit file, or a file permission that was configured through a machine config. This causes configuration drift. Configuration drift can cause problems between nodes in a Machine Config Pool or when the machine configs are updated.  
+
+The Machine Config Operator (MCO) uses the Machine Config Daemon (MCD) to check nodes for configuration drift on a regular basis. If detected, the MCO sets the node and the Machine Config Pool (MCP) to `Degraded` and reports the error. A degraded node is online and operational. However, it cannot be updated.
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:10]: Machine Config Pool -> machine config pool

File: modules/machine-config-drift-detection.adoc [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+The MCD performs configuration drift detection upon each of the following conditions:
+
+* When a node boots.
+* After any of the files (Ignition files and systemd drop-in units) specified in the machine config are modified outside of the machine config.
+* Before a new machine config is applied.
++
+[NOTE]
+====
+If you apply a new machine config to the nodes, the MCD temporarily shuts down configuration drift detection. This shut down is needed because the current machine config necessarily differs from the new machine config. After the new machine config is applied, the MCD restarts detecting configuration drift using the new machine config.
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:20]: * This shut down -> This shutdown
* I'm not sure what "necessarily" means in the second sentence. Is this sentence trying to say that the current machine config "might" differ from the new one"? Or that it does differ (in which case maybe "necessarily" isn't needed at all?)

File: modules/machine-config-drift-detection.adoc [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+====
+
+When performing configuration drift detection, the MCD validates that the file contents and permissions fully match what the currently-applied machine config specifies. Typically, the MCD detects configuration drift in less than a second after the detection is triggered. 
+
+If the MCD detects configuration drift, the MCD:
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:25]: Per ISG, consider making this intro a full sentence (like "the MCD performs the following tasks", for example)

File: modules/machine-config-drift-detection.adoc [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+* Emits an error to the console logs
+* Emits a Kubernetes event
+* Stops further detection on the node
+* Sets the node and MCP to `Degraded`
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:30]: Elsewhere in the PR, I see `degraded` lowercase. I think uppercase is right, but can you check and adjust whichever needs to be so that it is consistent?

File: modules/machine-config-drift-detection.adoc [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+You can check if you have a degraded node by listing the MCPs:
+
+[source,terminal]
+----
+$ oc get mcp worker
+----
+
+If you have a degraded MCP, the `DEGRADEDMACHINECOUNT` field is non-zero, similar to the following:
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:39]: Per ISG, "following" shouldn't be used by itself - instead a word should follow it. Like "similar to the following output" or something like that

File: modules/machine-config-drift-detection.adoc [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+.Sample output
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:41]: .Sample output -> .Example output

File: modules/machine-config-drift-detection.adoc [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+ ...
+    Last Transition Time:  2021-12-20T18:54:00Z
+    Message:               Node ci-ln-j4h8nkb-72292-pxqxz-worker-a-fjks4 is reporting: "content mismatch for file \"/etc/mco-test-file\"" <1>
+    Reason:                1 nodes are reporting degraded status on sync
+    Status:                True
+    Type:                  NodeDegraded <2>
+ ...
+----
+<1> The node that contains the configuration drift. And, the error message indicating that the contents of the `/etc/mco-test-file`, which was added by the machine config, has changed outside of the machine config.
+<2> The state of the node is `Degraded`.
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:67]: The actual type listed here is `NodeDegraded`. Should that be what it is in the description instead of `Degraded`?

File: modules/machine-config-drift-detection.adoc [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+----
+ ...
+    Last Transition Time:  2021-12-20T18:54:00Z
+    Message:               Node ci-ln-j4h8nkb-72292-pxqxz-worker-a-fjks4 is reporting: "content mismatch for file \"/etc/mco-test-file\"" <1>
+    Reason:                1 nodes are reporting degraded status on sync
+    Status:                True
+    Type:                  NodeDegraded <2>
+ ...
+----
+<1> The node that contains the configuration drift. And, the error message indicating that the contents of the `/etc/mco-test-file`, which was added by the machine config, has changed outside of the machine config.
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:66]: Suggest rewording a bit to not start a sentence with "And". Could you say something like "This message shows that a node's `/etc/mco-test-file` file, which was added by the machine config, has changed outside of the machine config." or "This message lists a node whose `/etc/mco-test-file` file, which was added by the machine config, has changed outside of the machine config." Or I'm sure there's a better alternative.

File: modules/machine-config-drift-detection.adoc [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+                      {"pd.csi.storage.gke.io":"projects/openshift-gce-devel-ci/zones/us-central1-a/instances/ci-ln-j4h8nkb-72292-pxqxz-worker-a-fjks4"}
+                    machine.openshift.io/machine: openshift-machine-api/ci-ln-j4h8nkb-72292-pxqxz-worker-a-fjks4 <1>
+                    machineconfiguration.openshift.io/controlPlaneTopology: HighlyAvailable
+                    machineconfiguration.openshift.io/currentConfig: rendered-worker-67bd55d0b02b0f659aef33680693a9f9
+                    machineconfiguration.openshift.io/desiredConfig: rendered-worker-67bd55d0b02b0f659aef33680693a9f9
+                    machineconfiguration.openshift.io/reason: content mismatch for file "/etc/mco-test-file" <2>
+                    machineconfiguration.openshift.io/state: Degraded <3>
+ ...
+----
+<1> The node that contains the configuration drift.
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:92]: Is this callout necessary, because you already knew which node it was, per the oc command above.

File: modules/machine-config-drift-detection.adoc [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+<2> The error message indicating that configuration drift was detected between the node and the listed machine config. Here the error message indicating that the contents of the `/etc/mco-test-file`, which was added by the machine config, has changed outside of the machine config.
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:93]: Here the error message indicating -> Here the error message indicates

File: modules/machine-config-drift-detection.adoc [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+<3> The state of the node is `Degraded`.
+
+You can correct configuration drift and return the node to the `Ready` state by performing one of the following remediations:
+
+* Ensure that the contents and file permissions of the files on the node match what is configured in the machine config. You can manually rewrite the file
+contents or change the file permissions.
+* Generate a link:https://access.redhat.com/solutions/5414371[force file] on the degraded node. The force file causes the MCD to bypass the usual configuration drift detection and re-applies the current machine config.
```
Comment (bergerhoffer) [inline modules/machine-config-drift-detection.adoc:100]: re-applies -> reapplies

File: modules/machine-config-overview.adoc [hunk 15]
```diff
@@ -66,6 +66,8 @@ The MCO is not the only Operator that can change operating system components on
 
 Tasks for the MCO configuration that can be done post-installation are included in the following procedures. See descriptions of {op-system} bare metal installation for system configuration tasks that must be done during or before {product-title} installation.
 
+There might be situations where the configuration on a node does not fully match what the currently-applied machine config specifies. This state is called _configuration drift_. The Machine Config Daemon (MCD) regularly checks the nodes for configuration drift. If the MCD detects configuration drift, the MCO marks the node `degraded` until an admin corrects the node configuration. A degraded node is online and operational. However, it cannot be updated. For more information on configuration drift, see _Understanding configuration drift detection_ below.
```
Comment (bergerhoffer) [inline modules/machine-config-overview.adoc:69]: I'd remove "below", because it might not necessarily be below.

File: modules/machine-config-drift-detection.adoc (hunk 5 above)
Comment (mburke5678) [reply modules/machine-config-drift-detection.adoc:20]: @bergerhoffer _This shutdown is needed because this new machine config necessarily differs from the machine config on the nodes._
If you apply a new machine config to nodes, there will be changes, and this new machine config is going to be different than the one currently on the nodes.
